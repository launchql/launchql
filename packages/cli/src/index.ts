#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});