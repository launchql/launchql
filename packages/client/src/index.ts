import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}