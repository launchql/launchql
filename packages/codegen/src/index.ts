import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
