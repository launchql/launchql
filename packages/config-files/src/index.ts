export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';