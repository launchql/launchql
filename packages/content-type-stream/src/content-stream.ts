// @ts-nocheck
import crypto from 'crypto';
import stream from 'stream';

// TODO move these into this monorepo
import uuid from 'uuid-hash';
import etag from 'etag-hash';

export class ContentStream extends stream.Transform {
  constructor() {
    super();
    this.shasum = crypto.createHash('sha1');
    this.etagsum = etag.createHash();
    this.uuidsum = uuid.createHash();
  }
  _write(chunk, enc, next) {
    this.shasum.update(chunk);
    this.etagsum.update(chunk);
    this.uuidsum.update(chunk);
    this.push(chunk);
    next();
  }
  _flush(done) {
    this.emit('contents', {
      uuid: this.uuidsum.digest(),
      sha: this.shasum.digest('hex'),
      etag: this.etagsum.digest()
    });
    done();
  }
}
