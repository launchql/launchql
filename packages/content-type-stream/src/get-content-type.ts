import mime from 'mime';
import { extname } from 'path';

const special: Record<string, Record<string, string>> = {
  ts: {
    ascii: 'text/x-typescript',
    binary: 'video/mp2t'
  },
  tsx: {
    ascii: 'text/x-typescript'
  }
};

const needsMimeReplacement: string[] = [
  'text/plain',
  'application/zip',
  'application/octet-stream',
  'application/vnd.ms-opentype',
  'application/font-sfnt'
];

const directReplacements: Record<string, string> = {
  'font/ttf': 'application/x-font-ttf',
  'application/vnd.ms-opentype': 'application/x-font-opentype',
  'font/otf': 'application/x-font-opentype',
  'font/woff2': 'application/font-woff2',
  'font/woff': 'application/font-woff',
  'font/eot': 'application/vnd.ms-fontobject',
  'image/svg': 'image/svg+xml'
};

// https: //stackoverflow.com/questions/2871655/proper-mime-type-for-fonts
// svg as "image/svg+xml"(W3C: August 2011)
// ttf as "application/x-font-ttf"(IANA: March 2013)
// or "application/x-font-truetype"
// otf as "application/x-font-opentype"(IANA: March 2013)
// woff as "application/font-woff"(IANA: January 2013)
// woff2 as "application/font-woff2"(W3C W. / E.Draft: May 2014 / March 2016)
// eot as "application/vnd.ms-fontobject"(IANA: December 2005)
// sfnt as "application/font-sfnt"(IANA: March 2013)

export const getContentType = (
  file: string,
  type: string,
  charset?: string
): string => {
  const ext = extname(file).replace(/./, '');

  let final = type;
  if (needsMimeReplacement.includes(type)) {
    // @ts-ignore
    const test = mime.getType(file);
    if (test) final = test;
  }

  if (Object.prototype.hasOwnProperty.call(special, ext)) {
    const key = /ascii/.test(charset || '') ? 'ascii' : (charset || '');
    if (special[ext][key]) {
      final = special[ext][key];
    }
  }

  if (Object.prototype.hasOwnProperty.call(directReplacements, final)) {
    return directReplacements[final];
  }

  return final;
};
