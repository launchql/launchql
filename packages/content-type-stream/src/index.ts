export * from './content-stream';
export * from './content-type-stream';