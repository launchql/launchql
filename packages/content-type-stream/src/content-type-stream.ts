import { FileTypeDetector, BufferPeekStream } from 'mime-bytes';
import type { Readable } from 'stream';

import { getContentType } from './get-content-type';

// Create a shared detector instance with default settings
const detector = new FileTypeDetector({
  peekBytes: 16384,
  checkMultipleOffsets: true
});

interface StreamContentTypeArgs {
  readStream: Readable;
  filename: string;
  peekBytes?: number;
}

interface StreamContentTypeResult {
  stream: BufferPeekStream | Readable;
  magic: { type: string; charset: string };
  contentType: string;
}

export async function streamContentType({
  readStream,
  filename,
  peekBytes = 16384
}: StreamContentTypeArgs): Promise<StreamContentTypeResult> {
  try {
    // Create a custom detector if different peek bytes are needed
    const customDetector = peekBytes !== 16384 
      ? new FileTypeDetector({ peekBytes, checkMultipleOffsets: true })
      : detector;

    // Detect from stream with fallback to filename
    const result = await customDetector.detectWithFallback(readStream, filename);
    
    // Get the peek stream that was created during detection
    // mime-bytes returns the peek stream as part of the result
    const peekStream = (result as any)?._stream || readStream;
    
    let type = 'application/octet-stream';
    let charset = 'binary';
    
    if (result) {
      type = result.mimeType;
      charset = result.charset || 'binary';
    }
    
    // Use existing content type resolution logic for special cases
    const contentType = getContentType(filename, type, charset);
    
    return {
      stream: peekStream,
      magic: { type, charset },
      contentType
    };
  } catch (error) {
    // If detection fails, return defaults with original stream
    const type = 'application/octet-stream';
    const charset = 'binary';
    const contentType = getContentType(filename, type, charset);
    
    return {
      stream: readStream,
      magic: { type, charset },
      contentType
    };
  }
}
