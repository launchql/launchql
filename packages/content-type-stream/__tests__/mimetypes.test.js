import { streamContentType } from '../src';
import { sync as glob } from 'glob';
import { createReadStream } from 'fs';
import { basename } from 'path';

const files = []
  .concat(glob(__dirname + '/../../../__fixtures__/kitchen-sink/**'))
  .concat(glob(__dirname + '/../../../__fixtures__/kitchen-sink/**/.*'))
  .filter((file) => {
    const key = file.split('kitchen-sink')[1];
    return key != '';
  })
  .map((f) => ({
    key: basename(f),
    path: f
  }));

const malicious = glob(__dirname + '/../../../__fixtures__/malicious/**')
  .filter((file) => {
    const key = file.split('malicious')[1];
    return key != '';
  })
  .map((f) => ({
    key: basename(f),
    path: f
  }));

describe('mimetypes', () => {
  it('good files', async () => {
    const res = {};
    const use = files;

    for (var i = 0; i < use.length; i++) {
      const file = use[i];
      const key = file.key;
      const readStream = createReadStream(file.path);
      const { stream: newStream, magic, contentType } = await streamContentType(
        {
          readStream,
          filename: file.path
        }
      );
      res[key] = {
        magic,
        contentType
      };
    }
    expect(res).toMatchSnapshot();
  });
  it('malicious files', async () => {
    const res = {};
    const use = malicious;

    for (var i = 0; i < use.length; i++) {
      const file = use[i];
      const key = file.key;
      const readStream = createReadStream(file.path);
      const { stream: newStream, magic, contentType } = await streamContentType(
        {
          readStream,
          filename: file.path
        }
      );
      res[key] = {
        magic,
        contentType
      };
    }
    expect(res).toMatchSnapshot();
  });
});
