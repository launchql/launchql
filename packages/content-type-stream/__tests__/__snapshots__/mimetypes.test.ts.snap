// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mimetypes good files 1`] = `
{
  ".gitignore": {
    "contentType": "text/plain",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  ".travis.yml": {
    "contentType": "text/yaml",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "LICENSE": {
    "contentType": "text/plain",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "ai.ai": {
    "contentType": "application/pdf",
    "magic": {
      "charset": "iso-8859-1",
      "type": "application/pdf",
    },
  },
  "apng.png": {
    "contentType": "image/png",
    "magic": {
      "charset": "binary",
      "type": "image/png",
    },
  },
  "bmp.bmp": {
    "contentType": "image/x-ms-bmp",
    "magic": {
      "charset": "binary",
      "type": "image/x-ms-bmp",
    },
  },
  "build.js": {
    "contentType": "application/javascript",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "build.js.map": {
    "contentType": "application/json",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "css.css": {
    "contentType": "text/css",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "csv.csv": {
    "contentType": "text/csv",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "docs.js": {
    "contentType": "application/javascript",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "docs.jsx": {
    "contentType": "text/jsx",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "docx.docx": {
    "contentType": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "magic": {
      "charset": "binary",
      "type": "application/zip",
    },
  },
  "dwg.dwg": {
    "contentType": "image/vnd.dwg",
    "magic": {
      "charset": "binary",
      "type": "image/vnd.dwg",
    },
  },
  "dxf.dxf": {
    "contentType": "image/vnd.dxf",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "emf.emf": {
    "contentType": "image/emf",
    "magic": {
      "charset": "binary",
      "type": "application/octet-stream",
    },
  },
  "epub.epub": {
    "contentType": "application/epub+zip",
    "magic": {
      "charset": "binary",
      "type": "application/epub+zip",
    },
  },
  "font.otf": {
    "contentType": "application/x-font-opentype",
    "magic": {
      "charset": "binary",
      "type": "application/vnd.ms-opentype",
    },
  },
  "font.ttf": {
    "contentType": "application/x-font-ttf",
    "magic": {
      "charset": "binary",
      "type": "application/font-sfnt",
    },
  },
  "font.woff": {
    "contentType": "application/font-woff",
    "magic": {
      "charset": "binary",
      "type": "application/octet-stream",
    },
  },
  "font.woff2": {
    "contentType": "application/font-woff2",
    "magic": {
      "charset": "binary",
      "type": "application/octet-stream",
    },
  },
  "gif-w-alpha.gif": {
    "contentType": "image/gif",
    "magic": {
      "charset": "binary",
      "type": "image/gif",
    },
  },
  "gif.gif": {
    "contentType": "image/gif",
    "magic": {
      "charset": "binary",
      "type": "image/gif",
    },
  },
  "jpg.jpg": {
    "contentType": "image/jpeg",
    "magic": {
      "charset": "binary",
      "type": "image/jpeg",
    },
  },
  "json.json": {
    "contentType": "application/json",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "less.less": {
    "contentType": "text/less",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "lock.lock": {
    "contentType": "text/plain",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "md.md": {
    "contentType": "text/markdown",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "mp4.mp4": {
    "contentType": "video/mp4",
    "magic": {
      "charset": "binary",
      "type": "video/mp4",
    },
  },
  "ods.ods": {
    "contentType": "application/vnd.oasis.opendocument.spreadsheet",
    "magic": {
      "charset": "binary",
      "type": "application/vnd.oasis.opendocument.spreadsheet",
    },
  },
  "odt.odt": {
    "contentType": "application/vnd.oasis.opendocument.text",
    "magic": {
      "charset": "binary",
      "type": "application/vnd.oasis.opendocument.text",
    },
  },
  "pct.pct": {
    "contentType": "image/x-pict",
    "magic": {
      "charset": "binary",
      "type": "application/octet-stream",
    },
  },
  "pdf.pdf": {
    "contentType": "application/pdf",
    "magic": {
      "charset": "binary",
      "type": "application/pdf",
    },
  },
  "png-w-alpha.png": {
    "contentType": "image/png",
    "magic": {
      "charset": "binary",
      "type": "image/png",
    },
  },
  "psd.psd": {
    "contentType": "image/vnd.adobe.photoshop",
    "magic": {
      "charset": "binary",
      "type": "image/vnd.adobe.photoshop",
    },
  },
  "rtf.rtf": {
    "contentType": "text/rtf",
    "magic": {
      "charset": "us-ascii",
      "type": "text/rtf",
    },
  },
  "scss.scss": {
    "contentType": "text/x-scss",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "shellscript": {
    "contentType": "text/x-shellscript",
    "magic": {
      "charset": "us-ascii",
      "type": "text/x-shellscript",
    },
  },
  "sql.sql": {
    "contentType": "application/x-sql",
    "magic": {
      "charset": "utf-8",
      "type": "text/plain",
    },
  },
  "svg-with-alpha-and-text.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "image/svg",
    },
  },
  "svg.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "image/svg",
    },
  },
  "svgo.yaml": {
    "contentType": "text/yaml",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "swf.swf": {
    "contentType": "application/x-shockwave-flash",
    "magic": {
      "charset": "binary",
      "type": "application/x-shockwave-flash",
    },
  },
  "tga.tga": {
    "contentType": "image/x-tga",
    "magic": {
      "charset": "binary",
      "type": "image/x-tga",
    },
  },
  "tif.tif": {
    "contentType": "image/tiff",
    "magic": {
      "charset": "binary",
      "type": "image/tiff",
    },
  },
  "todo-0.1.0.tgz": {
    "contentType": "application/x-gzip",
    "magic": {
      "charset": "binary",
      "type": "application/x-gzip",
    },
  },
  "tsv.tsv": {
    "contentType": "text/tab-separated-values",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "txt.txt": {
    "contentType": "text/plain",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "typescript.ts": {
    "contentType": "text/x-typescript",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "typescript.tsx": {
    "contentType": "text/x-typescript",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "wmf.wmf": {
    "contentType": "image/wmf",
    "magic": {
      "charset": "binary",
      "type": "application/octet-stream",
    },
  },
  "xlsx.xlsx": {
    "contentType": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "magic": {
      "charset": "binary",
      "type": "application/zip",
    },
  },
  "zip.zip": {
    "contentType": "application/zip",
    "magic": {
      "charset": "binary",
      "type": "application/zip",
    },
  },
}
`;

exports[`mimetypes malicious files 1`] = `
{
  "html-as-jpg.jpg": {
    "contentType": "text/html",
    "magic": {
      "charset": "us-ascii",
      "type": "text/html",
    },
  },
  "image-as-html.html": {
    "contentType": "image/jpeg",
    "magic": {
      "charset": "binary",
      "type": "image/jpeg",
    },
  },
  "js-as-image.jpg": {
    "contentType": "image/jpeg",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "js-as-svg.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
  "svg-with-script-tag-guya.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "image/svg+xml",
    },
  },
  "svg-with-script-tag.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "image/svg",
    },
  },
  "xml-as-xvg.svg": {
    "contentType": "image/svg+xml",
    "magic": {
      "charset": "us-ascii",
      "type": "text/plain",
    },
  },
}
`;
